### PR TITLE
Issue/128 povgen

### PR DESCRIPTION
**Pull request by vitalych**

### CODE REVIEW

File: src/BuildingS2E.rst [hunk 1]
```diff
@@ -169,8 +169,13 @@ Building the documentation
 --------------------------
 
 The S2E documentation is written in `reStructuredText <http://docutils.sourceforge.net/rst.html>`_ format. HTML
-documentation can be built using the S2E Makefile:
+documentation can be built as follows:
 
 .. code-block:: console
 
-    make -f $S2EDIR/Makefile docs
```
Comment (adrianherrera) [inline src/BuildingS2E.rst]: I thought you didn’t like the $ because you couldn’t copy and paste

File: src/Tutorials/PoV/index.rst [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+    $ s2e new_project -i debian-9.2.1-i386 -n vuln-lin32-32 --enable-pov-generation \
+      build/s2e/guest-tools32/common/demos/vulnerabilities @@
+
+Here is a breakdown of the ``s2e new_project`` command above:
+
+* It creates a new project called ``vuln-lin32-32`` that will run on the ``debian-9.2.1-i386`` image.
+
+* The ``--enable-pov-generation`` is important in order to generate PoVs. If you omit it, you will just get random test
+  cases, which will at best crash the binary depending on what the constraint solver generates. Vulnerability generation
+  involves a number of additional plugins that may add overhead to the analysis, so they are disable by default unless
```
Comment (adrianherrera) [inline src/Tutorials/PoV/index.rst]: Disabled

File: src/Tutorials/PoV/index.rst [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+.. note::
+
+    You should build the desired guest image first, or use the ``-d`` flag if you want to download a pre-built image
+    instead. Only Linux-based images are available for download. You must build Windows images yourself.
+
+3. Launch S2E and wait until it completes
+
+.. code-block:: console
+
+    $ cd ~/s2e/env/vuln-lin32-32
```
Comment (adrianherrera) [inline src/Tutorials/PoV/index.rst]: Should be s2e/env/projects

File: src/Tutorials/PoV/index.rst [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+    $ ./launch-s2e.sh
+
+Look at the files inside ``~/s2e/env/vuln-lin32-32/s2e-last``. This folder contains all the analysis results.
+You should have files that look like this:
+
+.. code-block:: console
+
+    $ ls ~/s2e/env/vuln-lin32-32/s2e-last
```
Comment (adrianherrera) [inline src/Tutorials/PoV/index.rst]: Add projects dir

File: src/Tutorials/PoV/index.rst [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+    [EIP+9] == $pc[3]
+    [EIP+10] == 0xff
+    [EIP+11] == 0xe
+
+Here is what every line of this recipe means:
+
+* ``reg_mask`` and ``pc_mask`` indicate which bits of the general purpose register and program counter can be
+  controlled by the attacker. It will be ``0xffffffff`` in almost all cases, meaning that all bits can be controlled.
+
+* ``type`` can be either 1 or 2 (see DARPA's CGC terminology). Type 1 recipes control register values. Type 2 allow
```
Comment (adrianherrera) [inline src/Tutorials/PoV/index.rst]: Do you have a link to the CGC documentation on this? Might be useful

File: src/Tutorials/PoV/index.rst [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+  In practice, if the guest OS uses data execution prevention (DEP)s, you will need to encode a ROP chain in the recipe.
+  The ``vulnerabilities`` demo allocates an executable area at a fixed location in order to simplify the recipe.
+
+
+
+PoVs for DARPA Decree/CGC binaries
+==================================
+
+DARPA's `Cyber Grand Challenge <http://archive.darpa.mil/cybergrandchallenge/>`_ (CGC) was the world's first all-machine
+hacking tournament. S2E was a key component in CodeJitsu's Cyber Reasoning System (CRS) and was used to automatically to
```
Comment (adrianherrera) [inline src/Tutorials/PoV/index.rst]: Just “to automatically “ (redundant “to”)

File: src/Tutorials/PoV/index.rst [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+
+
+**SeedSearcher**
+    Seed files (or test inputs) are concrete inputs for the program under analysis. These files can be anything that
+    the program accepts (PNG files, documents, etc.). They can be obtained from a fuzzer, generated by hand, etc.
+    For CGC, seeds are binary executables compiled from XML of C PoV format.
+
+    The ``SeedSearcher`` plugin fetches seed files to concolically guide execution in the target program. Seed files
+    are placed in the seeds directory.  During analysis, the ``SeedSearcher`` plugins polls the seeds directory for new
+    seeds.When it finds new seeds, the plugin forks a new state that fetches the new seed and then runs the binary
```
Comment (adrianherrera) [inline src/Tutorials/PoV/index.rst]: Space after “seeds.”

**Comment by adrianherrera** [comment]
Hey Vitaly, it looks really good! As well as the specific comments, one general comment is sticking to the 120 char line limit, which helps prevent line wrapping which is annoying to read (and I’m currently doing this on an iPad :)

File: src/BuildingS2E.rst (hunk 1 above)
Comment (vitalych) [reply src/BuildingS2E.rst]: That's true, especially for very long sequences of commands. I think in this case it's fine, and it also forces people to read what they copy/paste if they have to do one command at a time.

File: src/Tutorials/PoV/index.rst [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+    You should build the desired guest image first, or use the ``-d`` flag if you want to download a pre-built image
+    instead. Only Linux-based images are available for download. You must build Windows images yourself.
+
+3. Launch S2E and wait until it completes
+
+.. code-block:: console
+
+    $ cd ~/s2e/env/vuln-lin32-32
+    $ ./launch-s2e.sh
+
```
[the remark this answers is not in the document]
Comment (vitalych) [reply src/Tutorials/PoV/index.rst]: It should take a few seconds (less than 30 normally).

**Comment by vitalych** [comment]
Thanks for the review! It might also worth trying to do it from start to finish to get a sense of how long it will take for the audience.

**Comment by matrizzo** [comment]
I think it's good, assuming s2e and the images are already built, I think it took me about 20-30 minutes to go through the tutorial.

File: src/Tutorials/PoV/index.rst (hunk 2 above)
Comment (vitalych) [reply src/Tutorials/PoV/index.rst]: ok

File: src/Tutorials/PoV/index.rst [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+    $ cd ~/s2e/env
+    $ s2e new_project -i debian-9.2.1-i386 -n vuln-lin32-32 --enable-pov-generation \
+      build/s2e/guest-tools32/common/demos/vulnerabilities @@
+
+Here is a breakdown of the ``s2e new_project`` command above:
+
+* It creates a new project called ``vuln-lin32-32`` that will run on the ``debian-9.2.1-i386`` image.
+
+* The ``--enable-pov-generation`` is important in order to generate PoVs. If you omit it, you will just get random test
+  cases, which will at best crash the binary depending on what the constraint solver generates. Vulnerability generation
```
[the remark this answers is not in the document]
Comment (vitalych) [reply src/Tutorials/PoV/index.rst]: yes, I fixed most of them. Sometimes it's not possible, especially for links.

File: src/Tutorials/PoV/index.rst (hunk 3 above)
Comment (vitalych) [reply src/Tutorials/PoV/index.rst]: ok

File: src/Tutorials/PoV/index.rst (hunk 4 above)
Comment (vitalych) [reply src/Tutorials/PoV/index.rst]: ok

File: src/Tutorials/PoV/index.rst (hunk 5 above)
Comment (vitalych) [reply src/Tutorials/PoV/index.rst]: yes

File: src/Tutorials/PoV/index.rst (hunk 6 above)
Comment (vitalych) [reply src/Tutorials/PoV/index.rst]: ok

File: src/Tutorials/PoV/index.rst (hunk 7 above)
Comment (vitalych) [reply src/Tutorials/PoV/index.rst]: ok

**Comment by vitalych** [comment]
@matrizzo @adrianherrera I applied the feedback

**Pull request merged**